### ISSUE TITLE
View script variables without dll

### ISSUE THREAD

**Issue by nesrak1** [DerPopo/UABE#63]
If I'm correct, MonoBehaviour contains the values of the variables and things that a script uses. If I already know what the variables are in this specific script (two ints, one string), how do I view the values?

**Comment by DerPopo** [comment]
You can do such a thing with the type package and database editors (Options in the main window).
First, note the type of the MonoBehaviour (on the right when it's selected, e.g. 0xFFFFFFFF).
Open the .tpk file with the package editor and extract the type database that matches the file's Unity version best.
Then, using the database editor, add a new type, name it however you want and give it the type id you noted before. Press Edit and start adding the fields of any MonoBehaviour, then of the specific type :
Add the base field with Type=your class name, Name = Base, Size = -1, all boxes unchecked. Select the new base field.
Add PPtr<GameObject> m_GameObject with Size 12 (Unity 5, Unity 4 : 8) as a child. Add int m_FileID (size = 4) and SInt64 m_PathId (size = 8) for Unity 5 / int m_PathId (size = 4) for Unity 4 as childs. Select m_GameObject again, press Add (don't tick "Insert as child") and enter UInt8 m_Enabled (size = 1, align). Select the new field and add (not as child) PPtr<MonoScript> m_Script and add the same child fields as in PPtr<GameObject> before.
Add string m_Name (size = -1) after m_Script, then add Array Array (size = -1, Is Array and Align ticked) as a child, and int size (size=4) and char data (size=1) as children to Array.

It should look like this now, except the type name MonoBehaviour :
![monobehaviour](https://cloud.githubusercontent.com/assets/8328510/17434525/33ec7d0a-5b0b-11e6-9280-9e42275f1273.PNG)

Now, you can (finally) add the fields of the script. The string must look like m_Name, with size = -1 for variable size.

**Comment by 0x900x90** [comment]
You can export it as raw. All variables stored at the end. But you can't to know exactly which value represents specific variable. I edited script variable values. All worked fine. You just need to find right value :).

**Comment by nesrak1** [comment]
@DerPopo Every time I try to add something, the size just reverts back to -1
@0x900x90 I knew that already... If I wanted to do that I wouldn't have even asked the question, but I wanted an easier way to look at the variables without hex editing

**Comment by DerPopo** [comment]
It seems like the function I used (strtol) doesn't accept values with the sign bit set, so you'd need to type -10 or -0xA in your case instead. Already fixed it for the next release.
The always -1 size won't cause any issues though but it'll be fixed.

**Comment by nesrak1** [comment]
It took multiple attempts to get it to work (at first I realized it was actually 0xE6 not 0xF6, but that kept reverting). Once I finally did get it to actually stay at 0xE6, it just continued to delete the changes I was making (reverting to a default MonoBehaviour, exactly like the one in your screenshot except with the name "Knife"). I'm just clicking on string m_Name, clicking add, and adding two ints, a string, and a custom enum.

**Comment by nesrak1** [comment]
Am I doing something wrong? Am I not supposed to put the variables there? Or is this just a bug in UABE?
[I can find a screenshot if you need one]
[Also this is a bump]

**Comment by nesrak1** [comment]
Also in the API, how would I get the value of an enum?

**Comment by DerPopo** [comment]
The two number bugs (type id/size) will be fixed. It doesn't look like you're doing anything wrong but I couldn't reproduce the deleted changes.
Enum values are usually stored as ints.
Here's a sample type database with only the type you mentioned (for Unity 4 you'd have to change SInt64 to int) :
[MonoBehaviour test.zip](https://github.com/DerPopo/UABE/files/410599/MonoBehaviour.test.zip)

**Comment by nesrak1** [comment]
I did it exactly how you did and it worked.
Does this mean I'm required to put m_ at the beginning?

Also, I feel like there needs to be a copy paste function, because I have another script that involves lots and lots of strings and it's annoying trying to manually add each one.

**Comment by nesrak1** [comment]
What do I do with floats? Type float, length 4?

**Comment by DerPopo** [comment]
You don't have to put m_ at the beginning, you could also put any special characters in there, except if there are some ASCII function calls in between.
Floats are as you described. The size is unused at the moment, so it won't cause any issues if it's invalid.

**Comment by nesrak1** [comment]
Last thing, what do I do about MonoBehaviour? I tried <MonoBehaviour> but the results were all messed up. In the GameObject array, it uses Component?

**Comment by DerPopo** [comment]
Such numbers for PathID do exist in some Unity 5 files, especially bundles. I can't tell whether that's the case here.
If you defined all fields that don't have an exclusion attribute in the right order, which is not the alphabetical order that some decompilers display, and with the right type (PPtr in case of a Component such as Texture2D or MonoBehaviour; if it's a non-Component class/struct, add the child fields recursively, ...), the output should be correct. The PPtr<T> type used is not important.

**Comment by nesrak1** [comment]
Ended up being that there was a variable that I was supposed to mark private as public, so I'll see if this fixes anything. I'm only putting in the public variables into here right?

**Comment by DerPopo** [comment]
Every byte/UInt8 or UInt16 has alignment in MonoBehaviours if I remember correctly. Check if alignment is set (if you have such variables) and make sure all strings are aligned (check align for the Array).
In case you have any bundle files with such a MonoBehaviour inside, you can look if there are differences. Alignment is indicated with the numbers 0/1 in an exported dump.

**Comment by nesrak1** [comment]
![hereswhatitlookslike](https://cloud.githubusercontent.com/assets/12544505/17610560/caad3fb0-6004-11e6-9b96-c1d8331bca46.png)
Here's a real example from a real game.
They're all 0.

**Comment by DerPopo** [comment]
Could you upload a raw dump of the file?

**Comment by nesrak1** [comment]
[MonoBehaviour.zip](https://github.com/DerPopo/UABE/files/415682/MonoBehaviour.zip)

This?

**Comment by DerPopo** [comment]
In the file, the last 4*4 byte values match with what you have (damageType = 0, damage = 20.0, attackDist = 1.2, vfxEventType = 3), the three PPtrs also seem to be correct (m_ResourceClip has a 0 path id, which means resourceClip will be null; attackClip and miscClip point to fileId=0/pathId=164101).
But between m_Name and m_MiscClip, there definitely are other variables.
Probably they do come from a base class as you suspected.
By looking at the file, there seem to be at least two more PPtrs (both fileId = 0, one pathId=13624, other pathId=166357) and 3 floats (0.40, 0.33, 0.10) and some other values.

**Comment by nesrak1** [comment]
I tried putting in the base class variables, and only the first few worked. For some reason the bools were messing everything up, and I had to place useless UInt8s around the bools.
Now everything is finally working, sorry to bother so much.

**Comment by DerPopo** [comment]
Everything smaller than 4 bytes in a MonoBehaviour is aligned, so each bool takes 4 bytes. You can tick the alignment box in the editor for all bools/(U)Int8/(U)Int16 in a MonoBehaviour type.

**Comment by nesrak1** [comment]
Yep that fixed it. Thanks.

**Comment by nesrak1** [comment]
I'm going to need a way to either batch this, or copy and paste the same thing over and over.
I have 30 scripts with the same base class. The base class has about 15 variables. Will there be any way to use command line / API to add to this?

**Comment by DerPopo** [comment]
The API has a ClassDatabaseFile class. Basically you can call Read on an existing file and when the changes are done call Write, which will also update the header variables and the string table. Turning optimization off will speed it up greatly (the optimization is very slow, the next release will have an option for a faster linear one).
You can make a copy of the MonoBehaviour type (search for the right ClassDatabaseType using the class id) and insert it to the classes vector. Change the class id as you did in the editor and set the base class to the class id of MonoBehaviour (0x72). You can set the name by setting fromStringTable to false and then the string pointer in the union. 
The fields' variables are typeName (e.g. "int"), fieldName (e.g. "MyScript"), depth (= the amount of spaces in a dump, no need for recursion here), isArray/size (both as in the editor) and flags2 (0x4000 = alignment).

**Comment by nesrak1** [comment]
Does depth actually have anything to do with what's a child and what's not?
So Script Base is 0, and everything under it is 1?

**Comment by DerPopo** [comment]
The depth is indeed needed.
If a dump looks like this :
0SomeType Base
 0PPtr<SomeType> children
  0int m_FileID
  0SInt64 m_PathID
 1UInt8 alignMe
Base has depth = 0, children has 1, m_FileID and m_PathID have 2, alignMe has 1

**Comment by nesrak1** [comment]
What do I put for the write parameters?? I have
`classDatabaseFile->Write(AssetsWriterToFile, (LPARAM)pSaveFile, 0, false, 1Ul, true);`
But it's not writing anything... Its just overwrites the dat file with nothing.

**Comment by nesrak1** [comment]
Even trying to do the below code messes everything up.
```
classDatabaseFile->classes.push_back(classDatabaseFile->classes[findAssetClassByName(classDatabaseFile, "MonoBehaviour")]); //Make copy of monobehaviour
	classDatabaseFile->classes.back().classId = classID; //Set class ID to one we specified in parameters
	classDatabaseFile->classes.back().baseClass = 0x72;
	classDatabaseFile->classes.back().name.fromStringTable = false; //False so we can set name
	classDatabaseFile->classes.back().name.str.string = document["name"].GetString(); //Set name to name of class, specified by plugin
```

**Comment by DerPopo** [comment]
Here's a bit of (not perfect) sample code that works at least on my end (the only thing that might have changed is the order of parameters for ClassDatabaseFile::Write) :
```
ClassDatabaseFile cldbFile;
FILE *pReadFile = fopen("U5.4.0f3.dat", "rb");
cldbFile.Read(AssetsReaderFromFile, (LPARAM)pReadFile);
fclose(pReadFile);
for (size_t i = 0; i < cldbFile.classes.size(); i++)
{
	if (cldbFile.classes[i].classId == 0x72)
	{
		cldbFile.classes.push_back(cldbFile.classes[i]);
		cldbFile.classes.back().classId = -1;
		cldbFile.classes.back().baseClass = 0x72;
		cldbFile.classes.back().name.fromStringTable = false;
		cldbFile.classes.back().name.str.string = "MyOwnClass";
		FILE *pWriteFile = fopen("U5.4.0f3 Modded.dat", "wb");
		cldbFile.Write(AssetsWriterToFile, (LPARAM)pWriteFile, 0, 0, 1UL);
		fclose(pWriteFile);
		break;
	}
}
```
I don't know if that's the issue but if the file was opened in text mode, fwrite may interpret some of the data as non-Windows line-endings and "fix" them resulting in a broken binary file. For fopen, check that there's a b in the file mode, such as "rb"/"wb". Also, make sure the class database file is opened, not the package (.tpk).

**Comment by nesrak1** [comment]
Well, I created a for loop in another function which returned i. It didn't work, it gave me the same exact result that findAssetByClassName did. I then put the for loop in the actual code, and now it's working.
i is still exactly the same, so I'm not sure what's going on.

**Comment by nesrak1** [comment]
Last thing before I stop bothering you for good:
```
	classDatabaseFile->classes.back().fields.push_back(ClassDatabaseTypeField());
	classDatabaseFile->classes.back().fields.back().typeName.fromStringTable = false; //These weren't here before but they make no difference
	classDatabaseFile->classes.back().fields.back().typeName.str.string = "int"; //Type int
	classDatabaseFile->classes.back().fields.back().fieldName.fromStringTable = false;
	classDatabaseFile->classes.back().fields.back().fieldName.str.string = "m_test";
	classDatabaseFile->classes.back().fields.back().depth = 1; //This is 1, a parent variable
	classDatabaseFile->classes.back().fields.back().isArray = false; //Not an array
	classDatabaseFile->classes.back().fields.back().size = 4; //Int is 4 large
	classDatabaseFile->classes.back().fields.back().flags2 = 0x0000; //0 = no alignment
```
Is once again breaking the file.

**Comment by DerPopo** [comment]
I'll look into it asap but I'll finish bug hunting and "pressing the release button" before.

**Comment by nesrak1** [comment]
So nothing wrong with the code then? (Except the fact I'm lazily using pointers)

**Comment by nesrak1** [comment]
And when can we expect to see an API release?

**Comment by DerPopo** [comment]
It should be there in 1-2 days.

**Comment by nesrak1** [comment]
I just got the 2.0 API up and running, but it's still causing that exact same output.

**Comment by DerPopo** [comment]
I have copied your code into the example (replacing "classDatabaseFile->" by "cldbFile.") and everything seems to be fine.

**Comment by nesrak1** [comment]
It was just that fclose was commented out. It's not doing it anymore.

**Comment by nesrak1** [comment]
I'm sure this is just some c++ problem that I'm having but I can't get the PPtr around the names of anything as a typeName.str.string. If I try to concat using std::string and use.c_str(), it turns into question mark diamonds. If I try to strncpy/strncat, all of the PPtrs become the same thing. Putting a const char pointer without concat works fine.

**Comment by DerPopo** [comment]
In the first case, the std::string will be destroyed as soon as its scope ends with a `}` . The compiler automatically inserts a string destructor call there. The c_str return value will also be invalidated by that so classDatabaseFile->Write will use an invalid char pointer when called outside that scope.

If that's the reason it fails here, create std::strings in new memory instead and store the pointers in a list to delete them once the work is done :
```
std::vector<std::string*> buffers;
//Open a scope for any reason, such as a for-loop
{
    //const char *fieldTypeStr = ...
    //ClassDatabaseTypeField &currentField = classDatabaseFile->...
    std::string *pptrcat = new std::string("PPtr<") + fieldTypeStr + ">";
    buffers.push_back(pptrcat);
    const char* pptrcatchr = pptrcat.c_str();
    currentField.typeName.str.fromStringTable = false; //Also important, else it might interpret the string as a byte offset into the class database's string table.
    currentField.typeName.str.string = pptrcatchr;
}
classDatabaseFile->Write(/*insert parameters*/);
//Free the std::strings.
for (size_t i = 0; i < buffers.size(); i++)
    delete buffers[i];
//Optionally buffers.clear() but the scope thing also applies to vector of course.
```
Using std::vector<std::string> would cause issues when the vector's buffer is extended and the copy constructor on all elements needs to be called, potentially invalidating the char pointers. Generally, a push_back with a class type will also use the copy constructor to get the item into the vector.

Basically the same thing might be the reason for the char buffer method to fail. If you want to use it like in the example above, create a `std::vector<char*>` and use `char *pptrcatchr = new char[256];` as well as `delete[]` instead of `delete`.

**Comment by DerPopo** [comment]
Closing this since UABE 2.2 is able to retrieve the extended MonoBehaviour type information. Feel free to open a new issue in case you still have questions.
